Allow preloader to batch at deeper levels

jhawthorn opened the pull request:
Previously in #41385 we taught the preloader to batch similar queries together when possible, however (to reduce scope) we only did this for preloads at the top level of the provided Hash.

This commit adds a new private `Associations::Preloader::Batch` class which now handles the batching operation and can do so for loaders from different preloaders. This also modifies child_preloaders to return an un-executed preloader instead of running immediately, so that we can perform batching as we recurse each level.

One intentional limitation of this is that we're only attempting to batch together associations which are at the same depth in the provided Hash. This greatly simplifies the implementation and ensures a consistent grouping happens.

Another limitation (which I hope to address later) is that this still does not attempt to perform any batching of :through associations (however children of through associations may still be batched).

cc @dinahshi @eileencodes @seejohnrun @HParker 

---

For example, from the added test:

``` ruby
Author
  .where(name: "David")
  .includes(:thinking_posts => :comments, :welcome_posts => :comments)
  .first
```

**Queries before:**

``` sql
SELECT "authors".* FROM "authors" WHERE "authors"."name" = ? ORDER BY "authors"."id" ASC LIMIT ?
SELECT "posts".* FROM "posts" WHERE "posts"."title" = ? AND "posts"."author_id" = ?
SELECT "posts".* FROM "posts" WHERE "posts"."title" = ? AND "posts"."author_id" = ?
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = ?
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = ?
```

**Queries after:**

``` sql
SELECT "authors".* FROM "authors" WHERE "authors"."name" = ? ORDER BY "authors"."id" ASC LIMIT ?
SELECT "posts".* FROM "posts" WHERE "posts"."title" = ? AND "posts"."author_id" = ?
SELECT "posts".* FROM "posts" WHERE "posts"."title" = ? AND "posts"."author_id" = ?
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = ?
```

rafaelfranca commented on activerecord/lib/active_record/associations/preloader/batch.rb:
```diff
@@ -0,0 +1,36 @@
+          attr_reader :loaders
+
+          def group_and_load_similar(loaders)
+            loaders.grep_v(ThroughAssociation).group_by(&:grouping_key).each do |(_, _, association_key_name), similar_loaders|
```
this `grep_v` broke some tests in our app and we are currently trying to understand why N+1 queries are happening since the last change. @byroot has more info, but just wanted to give a heads up that the behavior of the preloader changed if you know why would do that.

jhawthorn committed:
Allow preloader to batch at deeper levels

Previously we taught the preloader to batch similar queries together
when possible, however (to reduce scope) we only did this for preloads
at the top level of the provided Hash.

This commit adds a new private Associations::Preloader::Batch class
which now handles the batching operation and can do so for loaders from
different preloaders. This also modifies child_preloaders to return an
un-executed preloader instead of running immediately, so that we can
perform batching as we recurse each level.

One intentional limitation of this is that we're only attempting to
batch together associations which are at the same depth in the provided
Hash. This greatly simplifies the implementation and ensures a
consistent grouping happens.

Another limitation (which I hope to address later) is that this still
does not attempt to perform any batching of :through associations
(however children of through associations may still be batched).

jhawthorn commented:
@rafaelfranca @byroot Rebased over main. I believe the fix in #41499 is still compatible with this.

rafaelfranca commented on activerecord/lib/active_record/associations/preloader.rb:
```diff
@@ -116,17 +116,23 @@ def preload(records, associations, preload_scope = nil)
 
-      private
-        attr_accessor :child_preloaders
+      attr_accessor :loaders
```
I think we should move this to be closer to the `attr_reader`
